jobs: TestRegistryLifecycle failed

[(jobs).TestRegistryLifecycle failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2837423&tab=buildLog) on [master@6bc054abae5d7e177666e7cc3bc736bee3a99d8b](https://github.com/cockroachdb/cockroach/commits/6bc054abae5d7e177666e7cc3bc736bee3a99d8b):

Fatal error:
```
panic: send on closed channel [recovered]
	panic: send on closed channel

```

Stack:
```
goroutine 46667 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc005860980, 0x51482e0, 0xc00019e010)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:233 +0x126
panic(0x3e71f00, 0x5060370)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql.(*syncIEResultChannel).unblockWriter(0xc0014c82a0, 0x5148360, 0xc0014c8330, 0x0, 0xc002158300, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal_result_channel.go:186 +0xfd
github.com/cockroachdb/cockroach/pkg/sql.(*syncIEResultChannel).nextResult(0xc0014c82a0, 0x5148360, 0xc0014c8330, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal_result_channel.go:204 +0x8c
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal(0xc0056fa480, 0x5148360, 0xc0014c8330, 0x44e04df, 0x1f, 0x516d260, 0xc0014c82a0, 0x0, 0x44839dc, 0x4, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:788 +0xd7b
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryIteratorEx(0xc0056fa480, 0x51482a0, 0xc005809340, 0x44e04df, 0x1f, 0x0, 0x44839dc, 0x4, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:549 +0x1d0
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).processClaimedJobs(0xc0019fb1e0, 0x51482a0, 0xc005809340, 0x50fb9e0, 0xc005bf4180, 0x13ee7d8, 0xc0010e1d20)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:73 +0x285
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).Start.func6(0x51482a0, 0xc005809340, 0x50fb9e0, 0xc005bf4180)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:705 +0x8e
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).Start.func1.1(0x51482a0, 0xc005809340)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:666 +0x282
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).Start.func10(0x51482e0, 0xc00019e010)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:811 +0x2c6
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc005860980, 0x51482e0, 0xc00019e010, 0x0, 0xc0058092c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:351 +0xb9
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:346 +0xfc

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestRegistryLifecycle/rollback
    jobs_test.go:205: Starting resume
    jobs_test.go:213: Exiting resume
    jobs_test.go:205: Starting resume
    jobs_test.go:213: Exiting resume
    jobs_test.go:255: Starting success
    jobs_test.go:259: Exiting success

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestRegistryLifecycle PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRegistryLifecycle.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/internal_result_channel.go
+++ b/pkg/sql/internal_result_channel.go
@@ -144,6 +144,11 @@ type syncIEResultChannel struct {
 	// goroutine exits its run() loop whereas waitCh is closed when closing the
 	// iterator.
 	dataCh chan ieIteratorResult
+
+	// waitCh is never closed. In all places where the caller may interact with it
+	// the doneCh is also used. This policy is in place to make it safe to unblock
+	// both the reader and the writer without any hazards of a blocked reader
+	// attempting to send on a closed channel.
 	waitCh chan struct{}
 
 	// doneCh is used to indicate that the ReadWriter has been closed.
@@ -195,7 +200,6 @@ func (i *syncIEResultChannel) unblockWriter(ctx context.Context) (done bool, err
 
 func (i *syncIEResultChannel) finish() {
 	close(i.dataCh)
-	close(i.waitCh)
 }
 
 func (i *syncIEResultChannel) nextResult(

